### PR TITLE
Add flag to skip MSBuild step during restore

### LINKED ISSUE #7690: NuGet restore for large solution is much slower in 4.8.1 vs. 3.5.0 due to MSBuild step
## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): NuGet.exe

NuGet version (x.x.x.xxx): 4.8.1.5435

dotnet.exe --version (if appropriate): N/A

VS version (if appropriate): 2017

OS version (i.e. win10 v1607 (14393.321)): win10 v1803 (17134.472)

Worked before? If so, with which NuGet version: 3.5.0.1996

## Detailed repro steps so we can see the same problem

In current versions of NuGet.exe, a restore operation with nothing to do (all packages are already restored) checks both the old-style packages.config files and the new PackageReferences. The enumeration of PackageReferences requires an MSBuild operation that was not required for the packages.config enumeration. For our large solution (400+ assemblies) that only uses packages.config files, a "no-op" restore takes about 17 seconds in NuGet 4.8.1, vs. under 1 second in 3.5.0.

I have a pull request I'll link shortly that adds a flag to skip the MSBuild step for projects like ours that don't need it.

### Verbose Logs

No interesting errors in the log. You just see the msbuild command start, and 17 seconds later finish.

### Sample Project

Only very large solutions reproduce the issue, so attaching an example is somewhat impractical.

### PR DESCRIPTION
**Pull request by ScotSalmon**
## Bug

Fixes: NuGet/Home#7690
Regression: Yes
* Last working version: 3.5.0 (as tested)
* How are we preventing it in future: optional flag with unit test

## Fix

Details: The MSBuild analysis of the projects is expensive on large projects (17 seconds on our 400-project solution). For legacy solutions (like ours) where all package references are handled in the old-style packages.config files, the MSBuild analysis is unnecessary and just wastes time on every package restore, even those that should be no-ops. Allow callers to pass a flag that skips this step.

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  N/A
Validation:  Ran all tests in NuGetRestoreCommandTests class and the three RestoreCommandTests classes. Observed 3 failures (RestoreCommand_LongPathPackage, Restore_LegacyPackageReference_WithNuGetLockFile, and Restore_LegacyPackageReference_WithNuGetLockFilePath) also observed in current unmodified dev. Added additional test for this specific case to confirm intended behavior with and without new flag, both cases pass.

### CODE REVIEW

**Comment by jainaashish** [comment]
Thanks @ScotSalmon for the quick PR. But as I mentioned on the issue, this needs a little more discussion. Current PR is quick n easy but way hacky and I'd imagine this functionality should be out of box, meaning NuGet should check if there is any PackageReference based project in the current solution which will require MSBuild evaluation otherwise just skip that part. That way we wouldn't need any new switch and will just work out of the box.

**Comment by ScotSalmon** [comment]
That sounds better for sure. I'll have a look at implementing that instead. If you have any pointers where to start, let me know.

**Comment by ScotSalmon** [comment]
@jainaashish let me know what you think of ef5433c.

Edit: nevermind; @alejandro5042 pointed out this will miss package references that come in through targets files. I can't think of how to do that without the full msbuild parse that I'm trying to avoid. I'd like to push for more consideration of the explicit switch.

(Also, I know my assumption that the XML PackageReference tag immediately follows the open angle bracket is unsafe, but compared to the targets file issue, that's easily fixable.)

**Comment by alejandro5042** [comment]
> As a quick check, what we can do is check the other way around, meaning check for if all the projects in the solution has packages.config file through existing code. 

Can a `.csproj` have a `packages.config` next to it but actually use `<PackageReference>`?  My manual testing suggests this is true. This might break backward compatibility if it suddenly didn't restore from `<PackageReference>`.

This can happen, for example, if:

  - A project that was upgraded to `<PackageReference>` but its `packages.config` was not removed
  - Multiple `.csproj` files in the same directory, of which only one of them uses the `packages.config` file (let's say they didn't use `packages.ProjectName.config`)

Another approach would be to introduce this flag to `nuget.config`. Perhaps named `<RestoreProjectStyle>`, similar to the [msbuild property](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master/docs/consume-packages/Package-References-in-Project-Files.md#using-packagereference-for-a-project-with-no-packagereferences).

- PRO: It saves the work of checking
- PRO: It does not introduce a flag to the CLI that users have to remember to pass in
- PRO: It does not introduce a flag to the CLI that clogs up the usage help
- CON: It is still not automatic, since it is still a flag

**Comment by jainaashish** [comment]
> Can a .csproj have a packages.config next to it but actually use <PackageReference>? My manual testing suggests this is true. This might break backward compatibility if it suddenly didn't restore from <PackageReference>.

> Multiple .csproj files in the same directory, of which only one of them uses the packages.config file (let's say they didn't use packages.ProjectName.config)

These are already error cases. There should always be either `PackageReference` or `packages.config` to manage NuGet dependencies per project. But NuGet checks for PR first whcih wins over PC hence it just works. Although I'm not sure if it works for all cases or will continue to work in future as well. At some point in time, we'd like to check for such cases before hand and fail with appropriate error message so that there is no confusion.

There are multiple ways to solve this, which is why I said this should be discussed a little before jumping into the solution to come up with the right experience. Adding few more folks who can help derive this discussion or could have even more proposals.

@rrelyea @anangaur @nkolev92

**Comment by ScotSalmon** [comment]
FWIW, the idea of just checking for packages.config and assuming the project uses ProjectReference if it doesn't exist won't work for our use case. We have some projects in the solution that don't have any NuGet references at all.

**Comment by jainaashish** [comment]
yeah that's the down side, which is why only msbuild evaluation can help understand if the project has any `PackageReference` or not....

**Comment by ScotSalmon** [comment]
Given that only msbuild evaluation can prove if the solution uses `PackageReference`, and msbuild evaluation is slow, it seems like we're down to some sort of manual switch, either my command line flag or Alejandro's nuget.config property. I'm open to other ideas if they're out there but those are the most promising options I see so far. I'm fine with either and like the advantages of using nuget.config that Alejandro noted in his post above.

**Comment by nkolev92** [comment]
I'm not a fan of an incomplete fix like the one in the current PR. 
We cannot really do quick checks accurately enough to make it the default behavior. 

A switch like that does not belong in nuget.config. The only option I'd be open to is a manual switch, but I'm interested what @rrelyea and @anangaur think.

**Comment by ScotSalmon** [comment]
@nkolev92 please refer to the original commit in this PR for a version with a manual switch.

**Comment by ScotSalmon** [comment]
@nkolev92 any thoughts about the manual switch option? The slowdown we're seeing is pretty painful here.

**Comment by nkolev92** [comment]
@rrelyea @anangaur thoughts?

**Comment by anangaur** [comment]
External switch sounds better in this situation. Let’s discuss in person once just to be on the same page.

**Comment by alejandro5042** [comment]
Will Visual Studio pass the switch when restoring? One benefit of using the `nuget.config` file is that the setting can apply in every restore situation.

There are already developer experience flags in the config file; e.g. `disableSourceControlIntegration`, `packageSources`, HTTP proxy, etc. I see skipping msbuild as an optimization for large projects until they move to `PackageReference` (if they ever do).  It should apply in every restore situation, for all developers in the repo. In contrast, a command line switch needs to be remembered or encoded in a build/restore script.

Regardless, it is encouraging to hear progress to resolve this performance issue!

**Comment by nkolev92** [comment]
@alejandro5042 

VisualStudio and commandline restore work differently in the way that they discover the projects. 
Note that you can do NuGet restore in 4 ways:

* VS
* nuget.exe
* dotnet.exe (PackageReference only)
* msbuild.exe (PackageReference only)

The core restore logic is the same, but in Visual Studio, NuGet gets it's information from the project system rather than run it's own msbuild evaluation. 

Basically, I think the commandline switch would only be relevant with nuget.exe.

**Comment by ScotSalmon** [comment]
@nkolev92 and all -- I just re-pushed the original version of my branch which uses the command line switch. Can we agree on that option?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs [hunk 1]
```diff
@@ -56,6 +56,9 @@ public class RestoreCommand : DownloadCommandBase
         [Option(typeof(NuGetCommand), "ForceRestoreCommand")]
         public bool Force { get; set; }
 
+        [Option(typeof(NuGetCommand), "RestoreCommandSkipMSBuild")]
+        public bool SkipMSBuild { get; set; }
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs]: Naming this `PackagesConfigOnly` would be more clear in my opinion.

Without reading the description of the command it isn't obvious what role MSBuild plays in restore, and if other commands such as Pack which requires MSBuild would break later if restore was called with SkipMSBuild.

Comment (ScotSalmon) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs]: Good point. Unless someone else objects I'll make that change.

Comment (ScotSalmon) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs]: Done.

**Comment by ScotSalmon** [comment]
@rrelyea @anangaur I think we're just waiting for your review here.

**Comment by zivkan** [comment]
Lock files for packages.config projects depends on the msbuild evaluation to discover which projects have lock files enabled and what the custom lock file name is, in case it's not using the default value. This change would therefore in effect also disable lock file usage. This needs to be taken into consideration (maybe it's just something that we need to document)

**Comment by ScotSalmon** [comment]
> This change would ... in effect also disable lock file usage. This needs to be taken into consideration (maybe it's just something that we need to document)

Would it be sufficient to mention that in the argument docs at https://github.com/NuGet/NuGet.Client/pull/2680/commits/060f2891930e33f70f6b99e133286bc8ea2309e5#diff-a15bcad9f22860661028a91cf708be67R5530? I could add "Using this flag disables lock file usage" to the end.

**Comment by zivkan** [comment]
Lock file for packages.config is still under development, so the message shouldn't be added until the feature is complete. It's just something that @rrelyea and @anangaur should consider when deciding if this PR should be approved/accepted.

**Comment by ScotSalmon** [comment]
@rrelyea @anangaur any updates from your side? It looks like we're back to just waiting for your review here.

**Comment by anangaur** [comment]
@ScotSalmon As I mentioned before, I am fine with the external switch for now but I need to have some discussion within the team before approving this PR - I want to understand the repercussions of supporting this option going forward. Ideally this seems like something that should be out of the box in NuGet.exe. But as per discussions above it doesn't seem to be possible without a MSBuild evaluation. So let me discuss with more folks in the team on the feasibility of an out-of-box solution. Will soon get back to you. Thanks for your continued patience :)

**Comment by ScotSalmon** [comment]
I haven't been maintaining this branch and we've changed our use case so this fix won't work for us anyway, so I'm dropping the PR to reduce clutter.
